### DISCUSSION TITLE
When I use foreach for regex.matches, var thinks its type is object. Why?

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/runtime#74919] [Q&A] [answered]
It can even use generic LINQ

**Comment by gfoidl** [comment, marked as answer]
It's because of https://github.com/dotnet/runtime/blob/62948580b4d3547473cb567bff6cc61308136d81/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MatchCollection.cs#L71-L73 and that C# binds to the instance method.
When you cast to `IEnumerable<Match>` first, then it bind to the explicit interface method.
```c#
foreach(var m in (IEnumerable<Match>)matches)
{
    // m is of type Match
}
```
Or if you use the explicit type instead of `var`...

I guess this is for historic reasons, as `MatchCollection` dates back to a time where generics weren't available.

**Comment by PathogenDavid** [comment]
As @gfoidl mentioned, `MatchCollection` predates generics so `foreach (var m in r.Matches("..."))` results in `m` being `object`. It's briefly mentioned here when `IEnumerable<T>` was added: https://github.com/dotnet/runtime/issues/13933

However, LINQ should pick up the explicit generic implementation as long as you're on a modern version of .NET. If you're still on .NET Framework you should be able to use `.Cast<Match>()` to use LINQ.

**Comment by zms9110750** [comment]
What happens if you delete the old API or change to the generic version of the iterator? How much will this affect the previous code?

**Reply by stephentoub** [reply]
That would break every compiled binary using the existing API.
